src/rdkafka_sasl_oauthbearer_oidc.c

```diff
@@ -0,0 +1,507 @@
+            strlen("Authorization: Basic ") + client_authorization_out.size + 1;
+        authorization_base64_header =
+            rd_malloc(authorization_base64_header_size);
+        rd_snprintf(authorization_base64_header,
+                    authorization_base64_header_size, "Authorization: Basic %s",
+                    client_authorization_out.ptr);
+
+        rd_free(client_authorization_in.ptr);
+        rd_free(client_authorization_out.ptr);
+        return authorization_base64_header;
+}
+
+
+/**
+ * @brief Build headers for HTTP(S) requests based on \p client_id
+ *        and \p client_secret. The result will be returned in \p *headersp.
+ *
+ * @locality Any thread.
+ */
+static void rd_kafka_oidc_build_headers(const char *client_id,
+                                        const char *client_secret,
+                                        struct curl_slist **headersp) {
+        char *authorization_base64_header;
+
+        authorization_base64_header =
+            rd_kafka_oidc_build_auth_header(client_id, client_secret);
+
+        *headersp = curl_slist_append(*headersp, "Accept: application/json");
+        *headersp = curl_slist_append(*headersp, authorization_base64_header);
+
+        *headersp = curl_slist_append(
+            *headersp, "Content-Type: application/x-www-form-urlencoded");
+
+        rd_free(authorization_base64_header);
+}
+
+/**
+ * @brief The format of JWT is Header.Payload.Signature.
+ *        Extract and decode payloads from JWT \p src.
+ *        The decoded payloads will be returned in \p *bufplainp.
+ *
+ * @returns Return error message while decoding the payload.
+ */
+static char *rd_kafka_jwt_b64_decode_payload(const char *src,
+                                             char **bufplainp) {
+        char *converted_src;
+        char *payload = NULL;
+        char *errstr  = NULL;
+
+        int i, padding, len;
+
+        int payload_len;
+        int nbytesdecoded;
+
+        int payloads_start = 0;
+        int payloads_end   = 0;
+
+        len           = (int)strlen(src);
+        converted_src = rd_malloc(len + 4);
+
+        for (i = 0; i < len; i++) {
+                switch (src[i]) {
+                case '-':
+                        converted_src[i] = '+';
+                        break;
+
+                case '_':
+                        converted_src[i] = '/';
+                        break;
+
+                case '.':
+                        if (payloads_start == 0)
+                                payloads_start = i + 1;
+                        else {
+                                if (payloads_end > 0) {
+                                        errstr =
+                                            "The token is invalid with more "
+                                            "than 2 delimiters";
+                                        goto done;
+                                }
+                                payloads_end = i;
+                        }
+
```

Comment:
Is there a missing break here?
If not, then add a comment like so:
```c
  /*FALLTHRU*/
```

to indicate that the implicit fallthru is intentional.

Reply:
Yeah, this fallthru is intentional. I still want to copy the `.` to the tmp token holder converted_src.